### PR TITLE
[internal] use webpack for builds

### LINKED ISSUE #4758: [internal] use webpack for building desktop-gui, reporter, runner
Purpose:
- to improve the development experience for Cypress contributors, speed up build/watch times for sub packages `desktop-gui`, `reporter`, `runner`.

### PR DESCRIPTION
**Pull request by kuceb**
**This PR is to improve the development experience for contributors to Cypress, and should only have minor changes for Cypress users**

fix #4758 
#### features:
- handles js, ts, coffeescript, jsx/tsx, and scss
- runner with < 1s watch time, ~10 sec build time
- easy to enable sourcemaps
- chrome devtool sources are mapped to relative files via sourceURL, even without sourcemaps
- auto-refresh browser on changes to reporter/runner/driver (optional) 

#### changes:
**build script**:
- fix issue with dangling promise during copyPackages step
- remove use of symlinks / proxy package.jsons in final build, instead statically grep and transform require statements for symlinked packages, e.g.: `require("@packages/socket/lib/server")` -> `require("../../socket/lib/server")`
- remove unused `packages/reporter/dist`(unused, `4Mb`)
#### tasks:
- [x] set up build/watch tasks for runner/reporter
- [x] use webpack-dev-server for virtual fs (faster watch) and auto-refresh 
- [x] split reporter into separate bundle (for testing)
- [x] get reporter tests passing
- [x] random failing tests
- [x] remove use of browser field (use symlinks in project root)
- [x] change source to not assume browser field
- [x] random failing server/e2e tests
- [x] unit snapshot test for scripts/binary/util/packages
- [x] make build deps DRY, only one copy of babel deps, webpack, etc.
- [x] removed old build scripts / zunder
- [x] refactored webpack / babel configs to new package `packages/web-config`
- [x] use webpack to build `packages/desktop-gui`
- [x] add e2e smoke tests to ensure final prod build is what we expect
- [x] only run cross-origin script error-prone e2e tests in chrome (for now)

#### post-tasks
- [x] re-add circleci cache
- [x] remove logic building binary on this branch in `win-binary-build`

### CODE REVIEW

File: circle.yml [hunk 1]
```diff
@@ -65,39 +65,39 @@ jobs:
...
+          key: _v8-{{ arch }}-{{ .Branch }}-deps-server
       - restore_cache:
-          key: v5-{{ arch }}-{{ .Branch }}-deps-socket
+          key: _v5-{{ arch }}-{{ .Branch }}-deps-socket
       - restore_cache:
-          key: v5-{{ arch }}-{{ .Branch }}-deps-static
+          key: _v5-{{ arch }}-{{ .Branch }}-deps-static
       - restore_cache:
-          key: v5-{{ arch }}-{{ .Branch }}-deps-ts
+          key: _v5-{{ arch }}-{{ .Branch }}-deps-ts
```
Comment (Bkucera) [inline circle.yml]: add package.json hash for cache invalidation

**Review by chrisbreiding — CHANGES_REQUESTED** [review]
Ran into some issues while testing this out:

* In the terminal, there's no indication that the initial build is finished when I run `npm run watch`. This is more of a nice-to-have, but it's something I was used to with zunder where it would say `Finished building cypress_runner.js` and I'd know it was ready.
* When I initially ran a spec, everything worked fine, but after changing a file and having it live reload, it could no longer load the font files:

Screen Shot 2019-05-22 at 2 57 01 PM

File: packages/reporter/webpack.config.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
+import webpack, { Configuration, optimize, ResolvePlugin } from 'webpack'
```
Comment (chrisbreiding) [inline packages/reporter/webpack.config.ts]: Could we somehow abstract this configuration and the shared dependencies between the reporter and runner to make it easier to edit and upgrade the dependencies in the future?

**Review by chrisbreiding — CHANGES_REQUESTED** [review]

File: packages/reporter/webpack.config.ts (hunk 2 above)
Comment (Bkucera) [reply packages/reporter/webpack.config.ts]: that is probably a good idea

Comment (Bkucera) [reply packages/reporter/webpack.config.ts]: ok, how do you like this solution https://github.com/cypress-io/cypress/pull/4103/commits/cafedfae7c00370e81a4e24e767465cafea43265

**Comment by Bkucera** [comment]
> Ran into some issues while testing this out:
> 
> * In the terminal, there's no indication that the initial build is finished when I run `npm run watch`. This is more of a nice-to-have, but it's something I was used to with zunder where it would say `Finished building cypress_runner.js` and I'd know it was ready.
> * When I initially ran a spec, everything worked fine, but after changing a file and having it live reload, it could no longer load the font files:
> 
@chrisbreiding 
ok these should be fixed in https://github.com/cypress-io/cypress/pull/4103/commits/cafe52baf68e1356c9cf7447ba034a0dfde75003

File: packages/reporter/webpack.config.ts (hunk 2 above)
Comment (chrisbreiding) [reply packages/reporter/webpack.config.ts]: Much better! 👍

**Comment by chrisbreiding** [comment]
What about using [webpack-dashboard](https://github.com/FormidableLabs/webpack-dashboard)? That would further improve the CLI experience.

**Comment by Bkucera** [comment]
> What about using [webpack-dashboard](https://github.com/FormidableLabs/webpack-dashboard)? That would further improve the CLI experience.

@chrisbreiding 
Tried it out, I don't prefer it to the regular `webpack-dev-server` output. Stdout is harder to see and scrolling doesn't work very well for me

I added the `--progress` option to watch mode, which will give you this
image

and, when you edit and save `click.coffee`, you'll see this printed to stdout:

image

showing you what file was watched and what module was rebuilt.

**Comment by chrisbreiding** [comment]
Fair enough. I was wondering about the scrolling and thought it might be a problem.

**Review by chrisbreiding — DISMISSED** [review]

**Comment by Bkucera** [comment]
@brian-mann @bahmutov @chrisbreiding 
after I get your approvals I'll do the `post-tasks` and merge

**Comment by bahmutov** [comment]
I’m out for for next couple of days but I’m so impressed with the amount of work you put in already, Ben

**Comment by chrisbreiding** [comment]
Does this have [autoprefixer](https://github.com/postcss/autoprefixer) hooked up for the CSS?

File: packages/web-config/webpack.config.base.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+const env = process.env.NODE_ENV === 'production' ? 'production' : 'development'
+const args = process.argv.slice(2)
+const liveReloadEnabled = !args.includes('--no-livereload')
+const watchModeEnabled = args.includes('--watch') || args.includes('-w')
+// opt out of livereload with arg --no-livereload
+if (liveReloadEnabled && watchModeEnabled) console.log(chalk.gray(`\nLive Reloading is enabled. use ${chalk.bold('--no-livereload')} to disable`))
+process.env.NODE_ENV = env
+
+const config: webpack.Configuration = {
+	mode: 'none',
```
Comment (chrisbreiding) [inline packages/web-config/webpack.config.base.ts]: What's up with the spacing here and in other files? It should be 2-space indentation to remain consistent with the rest of the codebase.

Comment (Bkucera) [reply packages/web-config/webpack.config.base.ts]: We don't have a typescript linter yet, so could this wait for #4449 ?

**Comment by Bkucera** [comment]
added `autoprefixer`

File: packages/web-config/webpack.config.base.ts (hunk 3 above)
Comment (Bkucera) [reply packages/web-config/webpack.config.base.ts]: ah i see, didn't even notices I was using tabs. Fixed

**Comment by FrozenPandaz** [comment]
@Bkucera This is absolutely fantastic work :)

I just played around with it locally by adding the `webpack.config.base` to a project and noticed there's nothing to handle `tsconfig` paths. 

This plugin can be added to resolve Tsconfig paths: https://www.npmjs.com/package/tsconfig-paths-webpack-plugin

Here's an example project:
https://github.com/FrozenPandaz/demos/tree/cypress-poc-2
* The last commit fixes the issue for tsconfig paths.
* I'm not sure how one might pass in a `tsconfig.json` to use. Would it have to be through the `cypress.json`?

**Comment by Bkucera** [comment]
@FrozenPandaz sorry for the confusion, but this is to build cypress internally before we release the binary, not user's spec code. I would recommend against such a complex webpack configuration to build your spec code, you usually don't need much more than `babel` to bundle your spec files

**Comment by FrozenPandaz** [comment]
I see, I misunderstood. Thanks for the clarification!

**Review by brian-mann — CHANGES_REQUESTED** [review]
We need to skip the tests - or fail before trying to run them **if** you are attempting to run tests in development in electron that fail with eval source maps due to the `cypress://` protocol.

Another thought - run those tests in **both** chrome and electron, and **only** skip the tests in electron when in development. That makes it clear what's going on and doesn't hinder local development.

**Review by brian-mann — CHANGES_REQUESTED** [review]
- [x] create a build-prod `npm script` for each package that has a `build` script and set NODE_ENV=production
- [x] ensure that NODE_ENV=development is clearly set on all other build packages (error if NODE_ENV is not set) -- or if it makes more sense to default it, that's fine too
- [x] update the build binary scripts to automatically find packages with a `build` script and error if its missing `build-prod`
- [x] iterate through `chrome` and `electron` and use `process.versions.electron` to decide whether to run the failing `eval` tests

**Comment by Bkucera** [comment]
@bahmutov for some reason, the second time circle tries to build this branch, I get `rebuild-node-sass` not found in both desktop-gui and runner packages
```
[runner:prebuild     ] sh: 1: rebuild-node-sass: not found
[runner:prebuild     ] npm ERR! file sh
[runner:prebuild     ] npm ERR! code ELIFECYCLE
```
I'm guessing for some reason the circle cache is not restoring the `node_modules/.bin/rebuild-node-sass`, or that it is not being backed up to cache, perhaps being deleted before cache is saved.

But I'm not sure why its happening only on this branch and not `develop`

**Comment by bahmutov** [comment]
This branch is cursed

**Comment by brian-mann** [comment]
Did you reenable circle cache?

**Comment by brian-mann** [comment]
@Bkucera 	and @bahmutov 	btw if you look at the services `circle.yml` file here you can see what I did so as to not repeat the same `v12` circle cache number each time...

https://github.com/cypress-io/cypress-services/blob/develop/.circleci/config.yml#L133

This allows you to move that value to a env var that you can modify without needing to change any code - and bust everything at the same time.

**Comment by bahmutov** [comment]
well, we can just put this variable as env var in the `circle.yml` script - at least then it will be in a single spot

**Comment by Bkucera** [comment]
e2e test failing unrelated

**Review by brian-mann — APPROVED** [review]

**Pull request merged**